Fix LoadBalancer tests to be provider-agnostic

danwinship opened the pull request:
#### What this PR does / why we need it:
Drops all the provider checks from the LoadBalancer tests and instead just marks them all `[Feature:LoadBalancer]`. We will need corresponding updates in test-infra...

At the moment it doesn't make any attempt to distinguish between what different LoadBalancer-supporting platforms support (#123714). People who want to run them will have to just manually `--ginkgo.skip` the tests they can't pass for now.

(Note: originally this PR was about trying to get the LB tests to pass under cloud-provider-kind, but now it is just about re-enabling them on GCE.)

#### Which issue(s) this PR fixes:
Fixes #124338

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/kind cleanup
/sig network
/priority important-soon
/assign @aojea

danwinship committed:
Drop GCE-only tests and subtests from LoadBalancer tests

danwinship commented:
```
 Summarizing 7 Failures:
  [FAIL] [sig-network] LoadBalancers [Feature:LoadBalancer] [It] should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on the same nodes [sig-network, Feature:LoadBalancer]
  k8s.io/kubernetes/test/e2e/network/loadbalancer.go:866
  [FAIL] [sig-network] LoadBalancers [Feature:LoadBalancer] [It] should be able to change the type and ports of a UDP service [Slow] [sig-network, Feature:LoadBalancer, Slow]
  k8s.io/kubernetes/test/e2e/network/service.go:591
  [FAIL] [sig-network] LoadBalancers ExternalTrafficPolicy:Local [Feature:LoadBalancer] [Slow] [It] should work for type=LoadBalancer [sig-network, Feature:LoadBalancer, Slow]
  k8s.io/kubernetes/test/e2e/network/loadbalancer.go:990
  [FAIL] [sig-network] LoadBalancers [Feature:LoadBalancer] [It] should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on different nodes [sig-network, Feature:LoadBalancer]
  k8s.io/kubernetes/test/e2e/network/loadbalancer.go:736
  [FAIL] [sig-network] LoadBalancers [Feature:LoadBalancer] [It] should have session affinity work for LoadBalancer service with Cluster traffic policy [Slow] [LinuxOnly] [sig-network, Feature:LoadBalancer, Slow]
  k8s.io/kubernetes/test/e2e/network/service.go:266
  [FAIL] [sig-network] LoadBalancers [Feature:LoadBalancer] [It] should be able to switch session affinity for LoadBalancer service with Local traffic policy [Slow] [LinuxOnly] [sig-network, Feature:LoadBalancer, Slow]
  k8s.io/kubernetes/test/e2e/network/service.go:266
  [FAIL] [sig-network] LoadBalancers [Feature:LoadBalancer] [It] should be able to switch session affinity for LoadBalancer service with Cluster traffic policy [Slow] [LinuxOnly] [sig-network, Feature:LoadBalancer, Slow]
  k8s.io/kubernetes/test/e2e/network/service.go:266 
```

So all of the UDP LoadBalancer tests failed (which is a known bug, https://github.com/kubernetes-sigs/cloud-provider-kind/issues/58), and most of the SessionAffinity tests. ("should have session affinity work for LoadBalancer service with Local traffic policy" passed, but none of the others did.)

danwinship commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer
seeing if failures are consistent or flakes...

aojea commented on test/e2e/network/loadbalancer.go:
```diff
@@ -187,37 +173,8 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
 		// Change the services to LoadBalancer.
-
-		// Here we test that LoadBalancers can receive static IP addresses.  This isn't
-		// necessary, but is an additional feature this monolithic test checks.
```
this is 😬  thanks for removing it

aojea commented on test/e2e/network/loadbalancer.go:
```diff
@@ -377,54 +310,12 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
 
 		// Change the services to LoadBalancer.
-
-		// Here we test that LoadBalancers can receive static IP addresses.  This isn't
```
this is clearly copy paste

danwinship commented on test/e2e/network/loadbalancer.go:
```diff
@@ -141,13 +133,7 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
-
 		loadBalancerLagTimeout := e2eservice.LoadBalancerLagTimeoutDefault
-		if framework.ProviderIs("aws") {
-			loadBalancerLagTimeout = e2eservice.LoadBalancerLagTimeoutAWS
```
hm... I guess we should probably use the larger value (TimeoutAWS) rather than the smaller one

aojea commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer

aojea commented:
great, we moved from
```
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on different nodes expand_more	3m18s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should have session affinity work for LoadBalancer service with Cluster traffic policy [Slow] [LinuxOnly] expand_more	10m47s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers ExternalTrafficPolicy:Local [Feature:LoadBalancer] [Slow] should work for type=LoadBalancer expand_more	1m20s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on the same nodes expand_more	2m12s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should be able to switch session affinity for LoadBalancer service with Local traffic policy [Slow] [LinuxOnly] expand_more	10m13s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should be able to change the type and ports of a UDP service [Slow] expand_more	3m46s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should be able to switch session affinity for LoadBalancer service with Cluster traffic policy [Slow] [LinuxOnly] expand_more
```

To only
```
Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should be able to change the type and ports of a UDP service [Slow] expand_more	16m19s
Kubernetes e2e suite: [It] [sig-network] LoadBalancers ExternalTrafficPolicy:Local [Feature:LoadBalancer] [Slow] should work for type=LoadBalancer expand_more
```

This will not pass as the loadbalancer is type proxy, so it will always replace the source ip, we may try to use something like this https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/ip_transparency#original-source-listener-filter but it may not work in all environments

> Kubernetes e2e suite: [It] [sig-network] LoadBalancers ExternalTrafficPolicy:Local [Feature:LoadBalancer] [Slow] should work for type=LoadBalancer expand_more	


The UDP one fails because is waiting for a rejection and we have discussed the same for TCP, is not clear we can impose this behavior for loadbalancers in general)

https://github.com/kubernetes/kubernetes/blob/4878acc5e58268c2c59054dc25c8f53c6a861501/test/e2e/network/service.go#L609-L621

> [FAILED] UDP service 172.18.0.6:81 not rejected: timed out waiting for the condition

aojea commented:
with 
```diff
diff --git a/test/e2e/network/loadbalancer.go b/test/e2e/network/loadbalancer.go
index e68e9153d11..08896bbecae 100644
--- a/test/e2e/network/loadbalancer.go
+++ b/test/e2e/network/loadbalancer.go
@@ -495,8 +495,8 @@ var _ = common.SIGDescribe("LoadBalancers", ginkgo.Serial, func() {
                err = udpJig.Scale(ctx, 0)
                framework.ExpectNoError(err)
 
-               ginkgo.By("looking for ICMP REJECT on the UDP service's LoadBalancer")
-               testRejectedUDP(udpIngressIP, svcPort, loadBalancerCreateTimeout)
+               ginkgo.By("UDP service's LoadBalancer should not be reachable")
+               testNotReachableUDP(udpIngressIP, svcPort, loadBalancerCreateTimeout)
 
                ginkgo.By("Scaling the pods to 1")
                err = udpJig.Scale(ctx, 1)
```

for the other test `should work for type=LoadBalancer` we should add a new Label as Feature:LoadbalancerL4 or Layer4 or something that indicates that the loadbalancer provisioned operates at L4

danwinship commented:
> This will not pass as the loadbalancer is type proxy

The e2e test should check the LoadBalancerStatus and belatedly skip the test in that case. (If either `ingress.IPMode == v1.LoadBalancerIPModeProxy` or `ingress.IP == ""`)

aojea commented:
great , those two tests that are failing now will be fixed once we make them aware of the proxy mode and we implement that in cloud-provider-kind

> Kubernetes e2e suite: [It] [sig-network] LoadBalancers ExternalTrafficPolicy:Local [Feature:LoadBalancer] [Slow] should work for type=LoadBalancer expand_more	1m32s
> Kubernetes e2e suite: [It] [sig-network] LoadBalancers ExternalTrafficPolicy:Local [Feature:LoadBalancer] [Slow] should handle updates to ExternalTrafficPolicy field expand_more

danwinship commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer

danwinship committed:
Fix references to "ESIPP" and "OnlyLocal" in network e2e tests

aojea commented on test/e2e/network/loadbalancer.go:
```diff
@@ -1284,6 +968,14 @@ var _ = common.SIGDescribe("LoadBalancers ExternalTrafficPolicy: Local", framewo
+		// "ExternalTrafficPolicy: Local" + "IPMode: Proxy".
+		// https://issues.k8s.io/123714
+		ingress := &svc.Status.LoadBalancer.Ingress[0]
+		if ingress.IP == "" || (ingress.IPMode != nil && *ingress.IPMode == v1.LoadBalancerIPModeProxy) {
```
Line 1028 should only do the check if is not proxy ... in that case I dont think we should skip the test though

aojea replied:
Line 1278 hsa another FIXME related

danwinship replied:
The NodePort test (at line 1028) isn't even using a LoadBalancer service. (Yeah, the test case is in the wrong file.) It's connecting from a pod to a nodeport, so that's just exercising kube-proxy behavior... which we know has to work because otherwise this test wouldn't pass in the gce tests! Not sure what's going on here.

The line 1278 FIXME doesn't affect our CI, but I believe that test case would fail in a cluster running Cilium, because it goes out of its way to preserve the source IP even when it "doesn't have to", https://docs.cilium.io/en/stable/network/kubernetes/kubeproxy-free/#client-source-ip-preservation

aojea commented:
> Kubernetes e2e suite: [It] [sig-network] LoadBalancers [Feature:LoadBalancer] should only allow access from service loadbalancer source ranges [Slow]

can it be possible that when setting the proxy mode we are affecting this behavior?

danwinship commented:
Another possibility for the flakes is that we're hitting previously-untested kindnet behavior? Like, maybe some of the tests work when the pods are all on the same worker, but not when they're on different workers?

aojea commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer

aojea commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer
5 failures now :/

dims commented:
@danwinship please re-add the entire `test/e2e/network/loadbalancer.go` back in this PR once https://github.com/kubernetes/kubernetes/pull/124519 lands. thanks!

danwinship committed:
Add back the previously-removed loadbalancer e2e tests

(No changes from before they were deleted, and thus they don't
actually compile because of missing provider code.)

danwinship committed:
Add feature.LoadBalancer and remove provider-based skips from LB tests

danwinship commented on test/e2e/network/loadbalancer.go:
```diff
@@ -926,7 +926,7 @@ var _ = common.SIGDescribe("LoadBalancers", feature.LoadBalancer, func() {
 })
 
-var _ = common.SIGDescribe("LoadBalancers ESIPP", feature.LoadBalancer, framework.WithSlow(), func() {
+var _ = common.SIGDescribe("LoadBalancers ExternalTrafficPolicy: Local", feature.LoadBalancer, framework.WithSlow(), func() {
```
(FTR I'm doing this now just to keep all (most?) of the renaming in one PR; I didn't want people updating their CI configs to skip all the "ESIPP" tests and then have us rename them later...)

aojea commented:
@danwinship linter error

```
{ ScriptError 
test/e2e/network/loadbalancer.go:1453:6: ST1003: var previousHttpErrors should be previousHTTPErrors (stylecheck)
	var previousHttpErrors uint64 = 0
	    ^
test/e2e/network/loadbalancer.go:1500:3: ST1003: var currentHttpErrors should be currentHTTPErrors (stylecheck)
		currentHttpErrors := atomic.LoadUint64(&httpErrors)
		^
test/e2e/network/loadbalancer.go:1504:3: ST1003: var partialHttpErrors should be partialHTTPErrors (stylecheck)
		partialHttpErrors := currentHttpErrors - previousHttpErrors
		^
```

danwinship committed:
Update the "new" code to pass current lint

aojea commented:
/test pull-kubernetes-e2e-gce-100-performance

danwinship commented:
/retest-required

aojea commented:
/test pull-kubernetes-e2e-gce-100-performance

danwinship commented:
@aojea what are you running the perf test for?
anyway, I think this is ready to merge?

aojea commented:
> @aojea what are you running the perf test for?

wrong ones sorry, I wanted to check https://github.com/kubernetes/test-infra/pull/32585

/lgtm
/approve

Let's iterate
